MVVM Toolkit vNext: source generators

Sergio0694 opened the pull request:
(🦙) Testing something for Discord Preview, please ignore this text at the top for now while we test it. Thanks! This is going to be great feature in the Toolkit, can't wait! (🦙)

<hr/>

## Contributes to #3872
<!-- Add the relevant issue number after the "#" mentioned above (for ex: Fixes #1234) which will automatically close the issue once the PR is merged. -->

<!-- Add a brief overview here of the feature/bug & fix. -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one or more that apply to this PR. -->

<!-- - Bugfix -->
- Feature
- Performance improvements
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->


## Changelog
Changes discussed/tracked in the linked issue.


## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] New major technical changes in the toolkit have or will be added to the [Wiki](https://github.com/windows-toolkit/WindowsCommunityToolkit/wiki) e.g. build changes, source generators, testing infrastructure, sample creation changes, etc...
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [X] Contains **NO** breaking changes

Youssef1313 commented on Microsoft.Toolkit.Mvvm.SourceGenerators/Extensions/MemberDeclarationSyntaxExtensions.cs:
```diff
@@ -0,0 +1,57 @@
+
+            if (index != -1)
+            {
+                return memberDeclaration.WithModifiers(memberDeclaration.Modifiers.RemoveAt(index));
```
This may not preserve trivia if the removed modifier has any. No idea if that's okay, but wanted to note.

Sergio0694 replied:
Ah, good catch. Yeah that's fine, the generation step doesn't care about trivia at all (also to make things simpler). When we actually generate code at the end we just call `NormalizeWhitespace()` anyway that will fix everything 🙂

Youssef1313 commented on Microsoft.Toolkit.Mvvm.SourceGenerators/Input/ICommandGenerator.cs:
```diff
@@ -0,0 +1,341 @@
+        {
+            string propertyName = methodSymbol.Name;
+
+            if (SymbolEqualityComparer.Default.Equals(methodSymbol.ReturnType, context.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task")!) &&
```
Why a suppression is needed? Equals parameters are null annotated

Sergio0694 replied:
Ah, you're right, thanks! Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3873/commits/e97a60b19fbaa61882995de5acd32a15a736ea98.
I think I had originally added that suppression due to missing annotations with some older preview version of the Roslyn NuGet packages, and tghen forgot to go back and remove that here after updating the references 😄

Sergio0694 commented:
@michael-hawker As per our offline discussion, I've renamed the `[AlsoNotifyFor]` attribute to `[AlsoNotifyChangeFor]` to minimize possible breaking changes for preview users, and I've rebased the PR, I'm marking this as ready for review then 😄

michael-hawker commented:
@Youssef1313 thanks for taking a look. Were you able to try them out too? Feel free to sign-off on the PR if you're all good. 🙂

michael-hawker commented on Microsoft.Toolkit.Mvvm.SourceGenerators/AnalyzerReleases.Unshipped.md:
```diff
@@ -0,0 +1,19 @@
+
+Rule ID | Category | Severity | Notes
+--------|----------|----------|-------
+MVVMTK0001 | Microsoft.Toolkit.Mvvm.SourceGenerators.INotifyPropertyChangedGenerator | Error | See https://aka.ms/mvvmtoolkit
```
When are these used? Are they actually errors? Should we create a specific aka.ms link for these specific warning/error codes and have that point to a doc which covers each one?

Sergio0694 replied:
These are reported by the source generators if anything invalid is detected, so it's a way to make the error reporting proactive and to result in build time errors instead of some weird runtime crash or some hard to understand error due to incorrect generated code resulting from wrong usage of some attribute or something. They're used eg. here:

https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/1b08e47a2314791356a6c8ed7fd19f7684b828af/Microsoft.Toolkit.Mvvm.SourceGenerators/ComponentModel/TransitiveMembersGenerator.cs#L58-L63

And yeah they're actual errors that will cause the build to fail until they're resolved 😄

Not sure whether we need a dedicated page given that they should already include enough info in their error message, but I suppose that's something we could put together if you think it'd be useful. Maybe just a single page with a paragraph for each error and a code snippet to show an example or something, and steps on how to resolve it. Thoughts?

michael-hawker replied:
Yeah, I think that'd be good (they can just all be on a single page). I created a https://aka.ms/mvvmtoolkit/error link for us to point to that page (for now just points to doc root).

michael-hawker replied:
@Sergio0694 want to update the links to the new aka.ms for future proofing? We can redirect it later wherever we need for this specific purpose without needing to make a code-update.

Sergio0694 replied:
Sure thing! Done in 806b6da6f5606fe5af6a3ab429b777db4331c59b 🙂

michael-hawker commented on Microsoft.Toolkit.Mvvm/ComponentModel/ObservableObject.cs:
```diff
@@ -7,6 +7,14 @@
 // This file is inspired from the MvvmLight library (lbugnion/MvvmLight),
 // more info in ThirdPartyNotices.txt in the root of the project.
 
+// ================================== NOTE ==================================
```
Guess there's no way around this as their structure is slightly different, eh?

Sergio0694 replied:
Yeah, unfortunately there's some differences over all the file that make it impossible to just reuse the same file - we'd have to manually parse it and then exclude all the unwanted members and individual statements, and that'd be more error prone and expensive to do at runtime than just making a simple trimmed down copy like this we can reuse as is 🙂

michael-hawker commented on Microsoft.Toolkit.Mvvm/ComponentModel/Attributes/INotifyPropertyChangedAttribute.cs:
```diff
@@ -0,0 +1,40 @@
+        /// the two <see cref="ObservableObject.OnPropertyChanged(PropertyChangedEventArgs)"/> overloads will be generated.
+        /// The default value is <see langword="true"/>.
+        /// </summary>
+        public bool IncludeAdditionalHelperMethods { get; set; } = true;
```
If this is true by default, how is it different than using the `ObservableObjectAttribute`? Is this needed?

Sergio0694 replied:
The difference when `IncludeAdditionalHelperMethods` is set to `true` is that `[INotifyPropertyChanged]` only implements `INotifyPropertyChanged`, whereas `ObservableObject` implements both that and `INotifyPropertyChanging`, which is not really necessary in some cases (eg. neither UWP nor WinUI 3 actually use that). This attribute is meant to offer a more compact and lightweight option for devs just interested in the `INPC` interface, plus it has the additional ability to skip the helpers and just generate the absolute bare minimum implementation support as well. Makes sense? 😄

michael-hawker commented on Microsoft.Toolkit.Mvvm/ComponentModel/Attributes/ObservablePropertyAttribute.cs:
```diff
@@ -0,0 +1,44 @@
+    /// partial class MyViewModel : ObservableObject
+    /// {
+    ///     [ObservableProperty]
+    ///     private string name;
```
What happens if I want something camel cased like `IsEnabled`, if my field is named `isenabled` won't it generate `Isenabled`? Should we have another example for this if I need to define the field as `isEnabled` which seems a bit odd too.

Sergio0694 replied:
The property name generation logic is defined here:

https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/16ba7cc0c64310d0e82d04239183eece01161e7a/Microsoft.Toolkit.Mvvm.SourceGenerators/ComponentModel/ObservablePropertyGenerator.cs#L446-L460

That is correct, if your field is named `isenabled` then the resulting property will indeed be called `Isenabled`, but then again, one should never be calling anything `isenabled` anyway because that's against the naming guidelines 😄
This logic supports all common naming patterns, so `_fieldName`, `m_fieldName` and `fieldName`.

michael-hawker replied:
[4th Bullet](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-properties) "Is", "Can", "Has" are common property name starters. So yes, it should be like `_isEnabled` I guess. I think we should at least have an example in the comment that shows the CamelCase scenario though, so that folks can get the properly cased Property name out of the generator.

Sergio0694 replied:
I would expect that people using `_snakeCase` would already have the right casing in place and it'd "just work", given that nobody in C# lang does ever really _not_ use upper casing to distinguish words in any given symbol name. That said sure, we could add some remarks in the XML docs to call out the currently supported formats, would that work? 😄

michael-hawker replied:
Yeah, sounds great! Thanks! I think XML doc with the handled scenarios, plus another example property with the different syntax showing two words for CamelCasing the final property would cover all the bases.

Sergio0694 replied:
Done in https://github.com/CommunityToolkit/WindowsCommunityToolkit/pull/3873/commits/37caf88b68463ebb23b67c2e5d3f095485c9ecc5, also rebased to fix conflicts 😄

michael-hawker commented on Microsoft.Toolkit.Mvvm/ComponentModel/Attributes/ObservablePropertyAttribute.cs:
```diff
@@ -0,0 +1,44 @@
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Field, AllowMultiple = false, Inherited = false)]
+    public sealed class ObservablePropertyAttribute : Attribute
+    {
```
Because we have the problem above, do we want to have an alternate optional property/constructor for this which allows you to provide a custom alias for the name of the resulting property. e.g.

```cs
[ObservableProperty("IsEnabled")]
private bool isenabled
````

Sergio0694 replied:
See answer above. I don't think this is necessary given that 🤔

michael-hawker replied:
What about if you want to upgrade existing code where the field name you're using to drive a property is named differently than the property you exposed?

```cs
    private int hitCounter;

   public int Count
   {
       get => hitCounter;
       set => SetProperty(ref hitCounter, value);
   }
```

Though I know they could just rename `hitCounter`, but maybe they have other internal counters and are distinguishing...

Sergio0694 replied:
I would expect people to just rename the private field here, yeah. I mean this would just be:

```csharp
[ObservableProperty]
private int count;
```

For other cases (which would be weird though) I'd expect users to just fallback into explicit setters for the property as usual. After all we're not trying to support every possible case ever (we couldn't anyway). I just really don't like having actual symbol names as a hardcoded string in an attribute, that feels really "non C#-ish" to me 😄

michael-hawker replied:
Sure, just wanted to call this out as a potential thing. It's always something we can add later since the default is to just use the provided field name. So no harm in leaving for now and getting more feedback at large once we ship it.

michael-hawker reviewed (commented):
Couple of quick questions after having fiddled with it a bit more.

michael-hawker commented:
Thanks @Sergio0694 looks good to me. I've been playing with them already, never going to go back! 🦙❤

@XAML-Knight think you'll find this interesting too. Feel free to give it a quick go around.

XAML-Knight reviewed (changes requested):
Impressive check-in - I found a couple of small items.

Sergio0694 commented:
Hey @XAML-Knight, thank you for having a look, glad you liked the PR! 😄

Let me give some context as to why I was using those hardcoded strings instead of using `typeof(T)` before, and in fact in https://github.com/CommunityToolkit/WindowsCommunityToolkit/pull/3873/commits/9283f2c7859164bba14eb479c99454f33b906625 I've removed those usages altogether and just switched to hardcoded strings everywhere for consistency. I've also talked with a Roslyn engineer to confirm this was a commonly used pattern on their end as well, for good measure 🙂

The short version of why those usages you've highlighted in your review comments weren't using `typeof(T)` is that those types were simply not available, so I couldn't have used `typeof(T)` even if I had wanted to (which I did, initially). Note that the MVVM Toolkit generator does _not_ have a dependency on the MVVM Toolkit, so it can't see any of those types directly. The ones I was using `typeof(T)` for were types that I had manually linked to the generator project to enable that, which I've also removed now. The reason why `typeof(T)` is technically not entirely correct is that you're essentially loading a type that's not the same type as the one actually used in consuming projects. Source generators run at build time and they're a totally separate assembly that's only loaded by Roslyn when building a project (so they're not referenced by consuming projects directly either), and because of this they can't access runtime information from consuming projects such as `Type` instances - the consuming project is just not an assembly that's loaded into memory and that is being executed at all at that point. This is the reason why generators only work with syntax trees and semantic information (symbols), as those are the only type of information that is available at that point in the toolchain. Even when we were using `typeof(T)` before to simplify things, we were actually not referencing the same type that consumers would've used, but just loading a `Type` instance from a source copy of that types that happened to be in the compiled in the generator assembly, which also happened to match the one actually present in the MVVM Toolkit and therefore using the same fully qualified name when final consumers would actually use it at runtime. But it was not technically the same type. Now instead we're just matching the actual well known type name, which is what Roslyn also does frequently as well (eg. [see here](https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs#L3847-L3873)).

Hope this helps! 😄

XAML-Knight commented:
Thanks for the detailed explanation, @Sergio0694 

PR **Approved**

Sergio0694 commented:
@XAML-Knight No worries, glad I could help! And yeah for sure, I did see that "where possible", just figured it would be useful to share some additional context in here instead of just saying "yeah it's not possible here", either for you in case you weren't super familiar with source generators yet (as they're pretty new, or more like, still in preview themselves ahah), as well as others possibly reading through this PR. Glad everything makes sense, thank you for the review! 🙌
